core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/HealthMonitor.scala

```diff
@@ -285,6 +298,34 @@ class HealthMonitor private (val slickDataSource: SlickDataSource,
     samDAO.getStatus()
   }
 
+  private def checkBPM: Future[SubsystemStatus] = {
+    logger.debug("Checking Billing Profile Manager...")
+    val status = billingProfileManagerDAO.getStatus()
+    val subSystems = Option(status.getSystems).map(_.values().asScala.toList).getOrElse(List())
+
+    val messages =
+      if (subSystems.isEmpty) None
+      else {
+        val errors = for {
+          subSystemStatuses <- Option(status.getSystems).toList
+          (subSystem, subSystemStatus) <- subSystemStatuses.asScala
+          message <- Option(subSystemStatus.getMessages).map(_.asScala).getOrElse(Seq("none"))
+        } yield s"$subSystem: (ok: ${subSystemStatus.isOk}, message: $message)"
+        Some(errors)
+      }
+    Future(SubsystemStatus(status.isOk, messages))
```

Comment:
Why are you repeating the transformation in the for-comprehension? 
```suggestion
    Future(
      SubsystemStatus(
        status.isOk,
        Option(status.getSystems).map { subSystemStatuses => 
          for {
            (subSystem, subSystemStatus) <- subSystemStatuses.asScala
            message <- Option(subSystemStatus.getMessages).map(_.asScala).getOrElse(Seq("none"))
          } yield s"$subSystem: (ok: ${subSystemStatus.isOk}, message: $message)"
        )
      )
    )
```

Reply:
To be honest, Yunyi's code makes more sense to me. Is it wrong? Though I don't understand in either case why we return OK if subsystems have errors.